compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/analyze.rs

```diff
@@ -1,38 +1,412 @@
+        self.operator()
+            .get_inputs_iter()
+            .collect::<Vec<_>>()
+            .into_iter()
+            .map(|n| self.dag.get_operator(n.id))
+    }
+
+    pub(crate) fn operator(&self) -> DagOperator<'a> {
+        self.dag.dag.get_operator(self.id)
+    }
+
+    #[allow(unused)]
+    pub(crate) fn partition(&self) -> &InstructionPartition {
+        &self.dag.partitions[self.id]
+    }
+
+    pub(crate) fn variance(&self) -> &OperatorVariance {
+        &self.dag.variances[self.id]
+    }
+}
+
+pub struct VariancedDagOperatorMut<'a> {
+    dag: &'a mut VariancedDag,
+    id: OperatorIndex,
+}
+
+impl<'a> VariancedDagOperatorMut<'a> {
+    fn get_inputs_iter(&self) -> impl Iterator<Item = VariancedDagOperator<'_>> {
+        self.operator()
+            .get_inputs_iter()
+            .collect::<Vec<_>>()
+            .into_iter()
+            .map(|n| self.dag.get_operator(n.id))
+    }
+
+    pub(crate) fn operator(&self) -> DagOperator<'_> {
+        self.dag.dag.get_operator(self.id)
+    }
+
+    pub(crate) fn partition(&self) -> &InstructionPartition {
+        &self.dag.partitions[self.id]
+    }
+
+    #[allow(unused)]
+    pub(crate) fn partition_mut(&mut self) -> &mut InstructionPartition {
+        &mut self.dag.partitions[self.id]
+    }
+
+    #[allow(unused)]
+    pub(crate) fn variance(&self) -> &OperatorVariance {
+        &self.dag.variances[self.id]
+    }
+
+    pub(crate) fn variance_mut(&mut self) -> &mut OperatorVariance {
+        &mut self.dag.variances[self.id]
+    }
+}
+
+#[allow(unused)]
+struct ComposedInput<'a> {
+    dag: &'a VariancedDag,
+    id: OperatorIndex,
+    from: Vec<OperatorIndex>,
+}
+
+struct ComposedInputMut<'a> {
+    dag: &'a mut VariancedDag,
+    id: OperatorIndex,
+    from: Vec<OperatorIndex>,
+}
+
+impl<'a> ComposedInputMut<'a> {
+    fn apply(self) {
+        let maxed_variance = self
+            .from
+            .into_iter()
+            .map(|id| self.dag.get_operator(id).variance().clone())
+            .reduce(|acc, var| acc.partition_wise_max(&var))
+            .unwrap();
+        let mut input = self.dag.get_operator_mut(self.id);
+        *(input.variance_mut()) = maxed_variance;
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct VariancedDag {
+    pub(crate) dag: Dag,
+    pub(crate) partitions: Partitions,
+    pub(crate) variances: Variances,
+}
+
+impl VariancedDag {
+    fn try_from_partitioned(partitioned: PartitionedDag) -> Result<Self> {
+        // We compute the initial variances with noise at input nodes and NANs everywhere
+        // else.
+        let variances = partitioned.get_initial_variances();
+        let PartitionedDag { dag, partitions } = partitioned;
+        let mut varianced = Self {
+            dag,
+            partitions,
+            variances,
+        };
+
+        // We loop, forwarding the noise, until it settles.
+        for forward_count in 0..MAX_FORWARDING {
```

Comment:
the longest cycle of input to input transitions.
twice the number of input + 1 ? (each forward connect at least on 1 input with another)

Reply:
Not sure about what you call input here. Could you detail ?